samples/net/mqtt_azure/src/main.c

```diff
@@ -0,0 +1,470 @@
+
+	client->password = &password;
+
+	username.utf8 = (u8_t *)CONFIG_SAMPLE_CLOUD_AZURE_USERNAME;
+	username.size = strlen(CONFIG_SAMPLE_CLOUD_AZURE_USERNAME);
+
+	client->user_name = &username;
+
+	client->protocol_version = MQTT_VERSION_3_1_1;
+
+	/* MQTT buffers configuration */
+	client->rx_buf = rx_buffer;
+	client->rx_buf_size = sizeof(rx_buffer);
+	client->tx_buf = tx_buffer;
+	client->tx_buf_size = sizeof(tx_buffer);
+
+	/* MQTT transport configuration */
+	client->transport.type = MQTT_TRANSPORT_SECURE;
+
+	tls_config = &client->transport.tls.config;
+
+	tls_config->peer_verify = 2;
+	tls_config->cipher_list = NULL;
+	tls_config->sec_tag_list = m_sec_tags;
+	tls_config->sec_tag_count = ARRAY_SIZE(m_sec_tags);
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+	tls_config->hostname = TLS_SNI_HOSTNAME;
+#else
+	tls_config->hostname = NULL;
+#endif
+
+#if defined(CONFIG_SOCKS)
+	mqtt_client_set_proxy(client, &socks5_proxy,
+			      socks5_proxy.sa_family == AF_INET ?
+			      sizeof(struct sockaddr_in) :
+			      sizeof(struct sockaddr_in6));
+#endif
+}
+
+static void mqtt_event_handler(struct mqtt_client *const client,
+			       const struct mqtt_evt *evt)
+{
+	switch (evt->type) {
+	case MQTT_EVT_CONNACK:
+		if (evt->result) {
+			LOG_ERR("MQTT connect failed %d", evt->result);
+			break;
+		}
+
+		mqtt_connected = true;
+		LOG_DBG("MQTT client connected!");
+		break;
+
+	case MQTT_EVT_DISCONNECT:
+		LOG_DBG("MQTT client disconnected %d", evt->result);
+
+		mqtt_connected = false;
+		clear_fds();
+		break;
+
+	case MQTT_EVT_PUBACK:
+		if (evt->result) {
+			LOG_ERR("MQTT PUBACK error %d", evt->result);
+			break;
+		}
+
+		LOG_DBG("PUBACK packet id: %u\n", evt->param.puback.message_id);
+		break;
+
+	default:
+		LOG_DBG("Unhandled MQTT event %d", evt->type);
+		break;
+	}
+}
+
+static int publish(struct mqtt_client *client, enum mqtt_qos qos)
+{
+	char payload[] = "{id=123}";
+	char topic[] = "devices/" MQTT_CLIENTID "/messages/events/";
+	u8_t len = strlen(topic);
+	struct mqtt_publish_param param;
+
+	param.message.topic.qos = qos;
+	param.message.topic.topic.utf8 = (u8_t *)topic;
+	param.message.topic.topic.size = len;
+	param.message.payload.data = payload;
+	param.message.payload.len = strlen(payload);
+	param.message_id = sys_rand32_get();
+	param.dup_flag = 0U;
+	param.retain_flag = 0U;
+
+	return mqtt_publish(client, &param);
+}
+
+/* Random time between 10 - 15 seconds
+ * If you prefer to have this value more than CONFIG_MQTT_KEEPALIVE,
+ * then keep the application connection live by calling mqtt_live()
+ * in regular intervals.
+ */
+static u8_t timeout_for_publish(void)
+{
+	return (10 + sys_rand32_get() % 5);
+}
+
+static void publish_message(struct k_work *work)
+{
+	int rc;
+
+	if (!mqtt_connected) {
+		goto end;
+	}
+
+	rc = publish(&client_ctx, MQTT_QOS_1_AT_LEAST_ONCE);
+	if (rc) {
+		LOG_ERR("mqtt_publish ERROR");
+		goto end;
+	}
+
+	LOG_DBG("mqtt_publish OK");
+
+	rc = wait(APP_SLEEP_MSECS);
```

Comment:
This `wait` could still be called within a system queue, which is no good IMO (could block the queue).

I'd consider moving a call to `publish_message` into the main function, something like:
```
void main()
{
    ...

    err = connect_to_cloud_and_publish();
    if (err != 0) {
         return;    
    }

    while (true) {
        k_sem_take(&publish_sem);

        err = publish_message();
        if (err != 0) {
             break;
        }
    }
}
```

In such case, all that workqueue handler need to do is to feed the semaphore.

Reply:
done